File: go/2022/day08/day08.go
package main

import (
	"aoc/cast"
	"flag"
	"fmt"
	"os"
	"strings"
)

var inputFile = flag.String("inputFile", "input.txt", "Relative file path to use as input")

func main() {
	flag.Parse()
	bytes, err := os.ReadFile(*inputFile)
	if err != nil {
		panic("Unable to read the input file")
	}

	input := string(bytes)

	fmt.Printf("AoC 2022, Day8, Part1 solution is: %d \n", part1(input))
	fmt.Printf("AoC 2022, Day8, Part2 solution is: %d \n", part2(input))
}

func part1(input string) int {
	grid := parseInput(input)
	visibleCoords := map[[2]int]string{}

	for r := 1; r < len(grid)-1; r++ {
		// from left
		highestFromLeft := -1
		for c := 0; c < len(grid[0])-1; c++ {
			height := grid[r][c]
			if height > highestFromLeft {
				currentPoint := [2]int{r, c}
				visibleCoords[currentPoint] = "L"
				highestFromLeft = height
			}
		}

		// from right
		highestFromRight := -1
		for c := len(grid[0]) - 1; c > 0; c-- {
			height := grid[r][c]
			if height > highestFromRight {
				currentPoint := [2]int{r, c}
				visibleCoords[currentPoint] = "R"
				highestFromRight = height
			}
		}
	}

	for c := 1; c < len(grid[0])-1; c++ {
		// from top
		highestFromTop := -1
		for r := 0; r < len(grid)-1; r++ {
			height := grid[r][c]
			if height > highestFromTop {
				currentPoint := [2]int{r, c}
				visibleCoords[currentPoint] = "T"
				highestFromTop = height
			}
		}

		// from bottom
		highestFromBottom := -1
		for r := len(grid) - 1; r > 0; r-- {
			heigh := grid[r][c]
			if heigh > highestFromBottom {
				currentPoint := [2]int{r, c}
				visibleCoords[currentPoint] = "B"
				highestFromBottom = heigh
			}
		}
	}

	return len(visibleCoords) + 4
}

func part2(input string) int {
	grid := parseInput(input)

	bestScore := 0
	// iterate through every eligible tree
	for r := 1; r < len(grid)-1; r++ {
		for c := 1; c < len(grid[0])-1; c++ {
			score := visible(grid, r, c, -1, 0)
			score *= visible(grid, r, c, 1, 0)
			score *= visible(grid, r, c, 0, -1)
			score *= visible(grid, r, c, 0, 1)

			if score > bestScore {
				bestScore = score
			}
		}
	}

	return bestScore
}

func visible(grid [][]int, r, c, dr, dc int) int {
	count := 0
	startingHeight := grid[r][c]
	r += dr
	c += dc

	for r >= 0 && r < len(grid) && c >= 0 && c < len(grid[0]) {
		height := grid[r][c]
		if height < startingHeight {
			count++
		} else {
			count++
			break
		}

		r += dr
		c += dc
	}

	return count
}

func parseInput(input string) (ans [][]int) {
	lines := strings.Split(input, "\n")
	for _, line := range lines {
		var row []int
		for _, n := range strings.Split(line, "") {
			row = append(row, cast.ToInt(n))
		}
		ans = append(ans, row)
	}
	return ans
}

File: python/2022/day08/day08.py
import numpy as np


def part1(input: str) -> int:
    lines = parse_input(input)
    grid = transform_to_grid(lines)

    n = len(grid)
    m = len(grid[0])

    grid = np.array(grid)
    ans = 0

    for i in range(n):
        for j in range(m):
            h = grid[i, j]

            if j == 0 or np.amax(grid[i, :j]) < h:
                ans += 1
            elif j == m - 1 or np.amax(grid[i, (j + 1) :]) < h:
                ans += 1
            elif i == 0 or np.amax(grid[:i, j]) < h:
                ans += 1
            elif i == n - 1 or np.amax(grid[(i + 1) :, j]) < h:
                ans += 1

    return ans


def part2(input: str) -> int:
    lines = parse_input(input)
    grid = transform_to_grid(lines)

    n = len(grid)
    m = len(grid[0])

    grid = np.array(grid)
    ans = 0

    directions = [[0, 1], [0, -1], [1, 0], [-1, 0]]

    for i in range(n):
        for j in range(m):
            h = grid[i, j]
            score = 1

            for dir_i, dir_j in directions:
                new_i, new_j = i + dir_i, j + dir_j
                dist = 0

                while (0 <= new_i < n and 0 <= new_j < m) and grid[new_i, new_j] < h:
                    dist += 1
                    new_i, new_j = new_i + dir_i, new_j + dir_j

                    if (0 <= new_i < n and 0 <= new_j < m) and grid[new_i, new_j] >= h:
                        dist += 1

                score *= dist

            ans = max(ans, score)

    return ans


def parse_input(input: str) -> list[int]:
    with open(input) as fn:
        return fn.read().strip().split()


def transform_to_grid(lines: list[int]) -> list[int]:
    return [list(map(int, list(line))) for line in lines]


def main():
    puzzle_input_path = "2022/day08/input.txt"
    print(f"AoC2022 Day8, Part1 solution is: {part1(input=puzzle_input_path)}")
    print(f"AoC2022 Day8, Part2 solution is: {part2(input=puzzle_input_path)}")


if __name__ == "__main__":
    main()
